ecs-agent/api/ecs/client/ecs_client.go

```diff
@@ -0,0 +1,802 @@
+	if len(outpostARN) > 0 {
+		return []*ecsmodel.Attribute{
+			{
+				Name:  aws.String("ecs.outpost-arn"),
+				Value: aws.String(outpostARN),
+			},
+		}
+	}
+	return []*ecsmodel.Attribute{}
+}
+
+func (client *ecsClient) getCustomAttributes() []*ecsmodel.Attribute {
+	var attributes []*ecsmodel.Attribute
+	for attribute, value := range client.configAccessor.InstanceAttributes() {
+		attributes = append(attributes, &ecsmodel.Attribute{
+			Name:  aws.String(attribute),
+			Value: aws.String(value),
+		})
+	}
+	return attributes
+}
+
+func (client *ecsClient) SubmitTaskStateChange(change ecs.TaskStateChange) error {
+	if change.Attachment != nil && client.stscAttachmentCustomRetryBackoff != nil {
+		retryFunc := func() error {
+			err := client.submitTaskStateChange(change)
+			if err == nil {
+				return nil
+			}
+			return submitStateCustomRetriableError(err)
+		}
+		return client.stscAttachmentCustomRetryBackoff(retryFunc)
+	}
+	return client.submitTaskStateChange(change)
+}
+
+func (client *ecsClient) submitTaskStateChange(change ecs.TaskStateChange) error {
+	if change.Attachment != nil {
+		// Confirm attachment by submitting attachment state change via SubmitTaskStateChange API (specifically in
+		// the input's Attachments field).
+		var attachments []*ecsmodel.AttachmentStateChange
+		eniStatus := change.Attachment.Status.String()
+		attachments = []*ecsmodel.AttachmentStateChange{
+			{
+				AttachmentArn: aws.String(change.Attachment.AttachmentARN),
+				Status:        aws.String(eniStatus),
+			},
+		}
+
+		_, err := client.submitStateChangeClient.SubmitTaskStateChange(&ecsmodel.SubmitTaskStateChangeInput{
+			Cluster:     aws.String(client.configAccessor.Cluster()),
+			Task:        aws.String(change.TaskARN),
+			Attachments: attachments,
+		})
+		if err != nil {
+			logger.Warn("Could not submit task state change associated with confirming attachment",
+				logger.Fields{
+					field.Error:     err,
+					"attachmentARN": change.Attachment.AttachmentARN,
+					field.Status:    eniStatus,
+				})
+			return err
+		}
+
+		return nil
+	}
+
+	req := ecsmodel.SubmitTaskStateChangeInput{
+		Cluster:            aws.String(client.configAccessor.Cluster()),
+		Task:               aws.String(change.TaskARN),
+		Status:             aws.String(change.Status.BackendStatus()),
+		Reason:             aws.String(trimString(change.Reason, ecsMaxTaskReasonLength)),
+		PullStartedAt:      change.PullStartedAt,
+		PullStoppedAt:      change.PullStoppedAt,
+		ExecutionStoppedAt: change.ExecutionStoppedAt,
+		ManagedAgents:      change.ManagedAgents,
+		Containers:         formatContainers(change.Containers, client.shouldExcludeIPv6PortBinding, change.TaskARN),
+	}
+
+	_, err := client.submitStateChangeClient.SubmitTaskStateChange(&req)
+	if err != nil {
+		logger.Warn("Could not submit task state change", logger.Fields{
+			field.Error:       err,
+			"taskStateChange": change.String(),
```

Comment:
Should this be a codified field in the fields package?

Reply:
Synced offline, so far we have only been creating logger field constants for values that are called more than once. Since this one is only called once, we will leave as is.